core/state/db_state_writer.go

```diff
@@ -74,6 +75,15 @@ func (dsw *DbStateWriter) UpdateAccountData(address libcommon.Address, original,
 	if err := dsw.db.Put(kv.HashedAccounts, addrHash[:], value); err != nil {
 		return err
 	}
+
+	if account.Incarnation == 0 && original.Incarnation > 0 {
```

Comment:
why not `if account.Incarnation != original.Incarnation` ?

Reply:
I think that in that case `IncarnationMap` isn't too relevant. My understanding is that `IncarnationMap` is relevant when a contract is self-destructed and then recreated. Within a block, `IntraBlockState` rather than `IncarnationMap` is used to determine the right incarnation. But here we're treating a special corner case when at the end of the block a contract is destroyed and then re-recreated as a non-smart account (`account.Incarnation == 0`).

Reply:
Here the lines I'm adding to `UpdateAccountData` are copy-pasted from `DeleteAccount`.